GeneratorInterface/Pythia8Interface/plugins/Pythia8HepMC3Hadronizer.cc

```diff
@@ -0,0 +1,1089 @@
+    fJetMatchingHook->beforeHadronization(lheEvent());
+  }
+
+  bool py8next = fMasterGen->next();
+
+  double mergeweight = fMasterGen.get()->info.mergingWeightNLO();
+  if (fMergingHook.get()) {
+    mergeweight *= fMergingHook->getNormFactor();
+  }
+
+  //protect against 0-weight from ckkw or similar
+  if (!py8next || std::abs(mergeweight) == 0.) {
+    lheEvent()->count(lhef::LHERunInfo::kSelected, 1.0, mergeweight);
+    event3().reset();
+    return false;
+  }
+
+  if (fJetMatchingPy8InternalHook.get()) {
+    const std::vector<double> djrmatch = fJetMatchingPy8InternalHook->getDJR();
+    //cap size of djr vector to save storage space (keep only up to first 6 elements)
+    unsigned int ndjr = std::min(djrmatch.size(), std::vector<double>::size_type(6));
+    for (unsigned int idjr = 0; idjr < ndjr; ++idjr) {
+      DJR.push_back(djrmatch[idjr]);
+    }
+
+    nME = fJetMatchingPy8InternalHook->nMEpartons().first;
+    nMEFiltered = fJetMatchingPy8InternalHook->nMEpartons().second;
+  }
+
+  // update LHE matching statistics
+  //
+  lheEvent()->count(lhef::LHERunInfo::kAccepted, 1.0, mergeweight);
+
+  if (evtgenDecays.get())
+    evtgenDecays->decay();
+
+  event() = std::make_unique<HepMC::GenEvent>();
+  event3() = std::make_unique<HepMC3::GenEvent>();
+  bool py8hepmc = toHepMC.fill_next_event(*(fMasterGen.get()), event3().get());
+
+  if (!py8hepmc) {
+    return false;
+  }
+
+  // 0th weight is not filled anymore since v8.30x, pushback manually
+  event3()->weights().push_back(fMasterGen->info.weight());
+
+  //add ckkw/umeps/unlops merging weight
+  if (mergeweight != 1.) {
+    event3()->weights()[0] *= mergeweight;
+  }
+
+  if (fEmissionVetoHook.get()) {
+    nISRveto += fEmissionVetoHook->getNISRveto();
+    nFSRveto += fEmissionVetoHook->getNFSRveto();
+  }
+
+  // fill shower weights
+  // http://home.thep.lu.se/~torbjorn/pythia82html/Variations.html
+  if (fMasterGen->info.nWeights() > 1) {
+    for (int i = 0; i < fMasterGen->info.nWeights(); ++i) {
+      double wgt = fMasterGen->info.weight(i);
+      event3()->weights().push_back(wgt);
+    }
+  }
+
+  return true;
+}
+
+bool Pythia8HepMC3Hadronizer::residualDecay() {
+  Event *pythiaEvent = &(fMasterGen->event);
+
+  int NPartsBeforeDecays = pythiaEvent->size() - 1;  // do NOT count the very 1st "system" particle
+                                                     // in Pythia8::Event record; it does NOT even
+                                                     // get translated by the HepMCInterface to the
+                                                     // HepMC::GenEvent record !!!
+  //int NPartsAfterDecays = event().get()->particles_size();
+  int NPartsAfterDecays = 0;
+  for (auto p : (event3().get())->particles()) {
+    NPartsAfterDecays++;
+  }
+
+  if (NPartsAfterDecays == NPartsBeforeDecays)
+    return true;
+
+  bool result = true;
+
+  std::cout << "residualdecay in action " << NPartsBeforeDecays << " " << NPartsAfterDecays << std::endl;
```

Comment:
Please do not use thread unsafe `cout`. Use LogSomething instead (here and below, unless if screened by an appropriate verbosity flag, which I assume is only used for debug)